### ISSUE TITLE
Clean up critic bans with wrong names

### ISSUE THREAD

**Issue by theseion** [guillep/pharo-core#6246]
**Describe the problem**
The fix for issue #6244 means that banned rules may no longer be identified as such. This is the case for all rules for which the `#uniqueIdentifierName` differs from the class name:
```smalltalk
ReAbstractRule allSubclasses select: [ :class |
	class uniqueIdentifierName ~= class name ]
```

**Classes involved**
Manifest*

**Proposal**
These bans need to be cleaned up, i.e., the class names in the ban method literals need to be replaced.

Alternatively, we could implement a fallback but
1. that's ugly and
2. adds an overhead to every lint run (e.g. per method and class).

Maybe we also need both, as we can't clean up external packages.

**Version information:**
 - Version Pharo 9.0
 - Commit c82e3cc26b844b585dccb56874523a70d8608c78

**Expected development cost**
1 hour for the fallback, maybe half a day for writing a replacement script.

**Comment by Ducasse** [comment]
Do we backport it to pharo 80.

**Comment by MarcusDenker** [comment]
This only is needed due to using #uniqueIdentifierName instead of the name of the class.

So Pharo8 is ok, we just need to do it in Pharo9.

having a cleanup for the meta-data is needed anyway as I think we reference a lot of deleted methods there.

**Comment by theseion** [comment]
I agree.
Just a side note: it would be nice if we could define strategies to clean bans. My preferred strategy, for example, would be to reset the bans for a method whenever I change that method. Otherwise I won't even know that there's a banned rule there. And if I remove the part that was banned anyway there's no more need for the ban.

**Comment by dionisiydk** [comment]
I think we can do much better than current approach with magic methods on Manifest class:
- to disable critiques for given method just add pragma to the method
- to disable critiques for the class just add a method #disabledCritiques to the class (returning list of disabled rules)
- to disable critiques for the package add a method #disabledCritiques to manifest.

Nothing needs to be generated and parsed with this approach. All ban rules can be simply edited manually as normal code. Currently it is really a trick

**Comment by theseion** [comment]
That's true, although I don't like the pragmas that much (it's what we use now). Especially when you want to disable more than one rule it becomes messy quickly.
On the other hand, I find it great that pragmas allow you to add a comment as to why a rule was banned. That can help in many contexts.

How about putting that kind of information into class properties (or something like that, which would need to become part of the class definition somehow) and adding a view for editing that information, e.g. like the class comment editor? I'm thinking, why does the class comment have to be special? I may want to define arbitrary information on a class that should be part of the definition.

**Comment by dionisiydk** [comment]
That would return us back to original problem: how to disable rules for
specific methods and how to sync this information (rename method for
example)
Pragmas is straightforward and simplest approach to this. No need for any
special tools.

ср, 29 апр. 2020 г., 07:26 Max Leske <notifications@github.com>:

**Comment by MarcusDenker** [comment]
what I do not like with pragmas is that they then make a method that was simple look much more complex... imagine you have 5 lines of pragmas in front... not nice.

I would like to have a way to add meta data that can be shown when needed, not all the time.

**Comment by dionisiydk** [comment]
How often we really violate rules in our methods?

I doubt that small and simple method will have 5 broken rules.
But if a method really violates 5 rules or more what does it mean?
Maybe it is not bad that it looks ugly. It will give a feedback to the
developer that this method is intentionally special case for bad practice.
And I think the intention sign is what is really important to show in such
methods and not hide them.

From the other side how many critiques we "normaly" violate in real live?
I think it is not much than two per method.

So I do not think that your argument is a real issue.


ср, 29 апр. 2020 г., 09:50 Marcus Denker <notifications@github.com>:

**Comment by theseion** [comment]
> That would return us back to original problem: how to disable rules for
specific methods and how to sync this information (rename method for
example)
Pragmas is straightforward and simplest approach to this. No need for any
special tools.

That is true.

> So I do not think that your argument is a real issue.

I agree with that, as long as we are talking about active development. But in my case we are talking about systems were a lot of code was written before the introduction of lint rules. It's not always easy to refactor those legacy methods, but they work, so I wouldn't want to change them until I have a reason to. But until then I don't need them showing up in my lint rule runs.
Sure, hardly any will have more than 2 failing rules, but still.

**Comment by MarcusDenker** [comment]
We could just test it.

**Comment by dionisiydk** [comment]
ср, 29 апр. 2020 г. в 10:40, Max Leske <notifications@github.com>:

> That would return us back to original problem: how to disable rules for
> specific methods and how to sync this information (rename method for
> example)
> Pragmas is straightforward and simplest approach to this. No need for any
> special tools.
>
> That is true.
>
> So I do not think that your argument is a real issue.
> I agree with that, as long as we are talking about active development. But
> in my case we are talking about systems were a lot of code was written
> before the introduction of lint rules. It's not always easy to refactor
> those legacy methods, but they work, so I wouldn't want to change them
> until I have a reason to. But until then I don't need them showing up in my
> lint rule runs.
>
>
You can just disable all critiques for all such packages or for all legacy
classes.
Why it does not work?

For compatibility we can allow current approach as a special ban strategy
which user can define in own package manifests. It can be extracted to
external package and loaded when needed.

**Comment by theseion** [comment]
Because I don't want to deactivate the rule for the whole package or class. Only for legacy methods. When I add methods or classes to a that class / package I still want to see the critiques.

**Comment by dionisiydk** [comment]
Interesting.
Does it mean that your manifests "encode" all "bad" legacy class and method
names?
Did you accumulate them manually and slowly during your work?

ср, 29 апр. 2020 г. в 11:09, Max Leske <notifications@github.com>:

**Comment by theseion** [comment]
Interesting thought :) I suppose so, at least partially. There are of course some valid rule bans in new code too. Unfortunately, I'm talking about Pharo 1.3 so we're using pragmas to ignore lint rules. But I suppose collecting the pragmas would accomplish the same, at least on the method level. We've had to resort to other means to ban class rules (overrides and custom lint rule harness).

**Comment by theseion** [comment]
I just realized, I *am* in the process of migrating to Pharo 8 so actually most of those bans *will* be in manifests soon.

**Comment by Ducasse** [comment]
Sorry but I do not want pragma for the rule in the method bodies and I will not change my mind. No shit in the code!
We removed the coloring this is not to add multiple pragma.
I think that manifest = metadata and this is nice for that.
